ext/opcache/jit/zend_jit_x86.dasc

```diff
@@ -10221,7 +10221,7 @@ static int zend_jit_do_fcall(dasm_State **Dst, const zend_op *opline, const zend
 		}
 
 		if (ZEND_OBSERVER_ENABLED) {
-			|	SAVE_IP
+			|	SET_EX_OPLINE opline, r0
```

Comment:
Something looks wrong.

``FP`` at this point keeps the ``execute_data`` of the called function, but you pass ``opline`` of the caller function.
Right? Or may be I miss something?

I think it's better to keep ``SAVE_IP`` for cases when the previous code calculates the value in ``IP`` (we just can't always know the value of ``opline`` at compile-time).

Few observer related tests on ARM are failed.

Reply:
Yes, this is wrong. I tried it because you were suggesting _So for inlined trace functions we should load the EX(opline) by SET_EX_OPLINE opline instead of SAVE_IP._ I probably misunderstood what you meant.

However, I honestly don't understand the problem with https://github.com/php/php-src/commit/51c74a4fa96ffcd38d3147df9af24b8dcab25635. In general, this will go to `LOAD_IP_ADDR (func->op_array.opcodes + num_args)` which is later followed by `SAVE_IP`.

Which boils down to two instructions:
```
|| if (GCC_GLOBAL_REGS) {
|  LOAD_ADDR IP, addr <= mov reg, ((ptrdiff_t)addr)
|| } else {
|  ADDR_STORE aword EX->opline, addr, RX <= mov mem, ((ptrdiff_t)addr)
|| }
|| if (GCC_GLOBAL_REGS) {
|  mov aword EX->opline, IP
|| }
```

There are, to me, two choices, either make the code more complex by conditionally not doing the SAVE_IP and directly assigning `ADDR_STORE aword EX->opline, (func->op_array.opcodes + num_args), r0`. Or just do what I did in my original commit.

If you have a better alternative than 51c74a4fa96ffcd38d3147df9af24b8dcab25635, please feel welcome to propose one. At least I don't see one.

Reply:
> Yes, this is wrong. I tried it because you were suggesting _So for inlined trace functions we should load the EX(opline) by SET_EX_OPLINE opline instead of SAVE_IP._ I probably misunderstood what you meant.

I meant, it should be possible to extract the first ``opline`` of the function from the the trace and load it by  ``SET_EX_OPLINE``. Something like the following (I didn't check this):

```c
if (trac && (trace+1).op == ZEND_JIT_TRACE_VM) {
    SET_EX_OPINE (trace+1).opline, r0
} else {
    SAVE_IP
}
```

> However, I honestly don't understand the problem with https://github.com/php/php-src/commit/51c74a4fa96ffcd38d3147df9af24b8dcab25635. In general, this will go to LOAD_IP_ADDR (func->op_array.opcodes + num_args) which is later followed by SAVE_IP in the if (ZEND_OBSERVER_ENABLED) branch.

1. This will execute code that wasn't designed for tracng JIT. I'm not sure if it's going to be right or not.
2. In any case, this code is going to do extra work (calculation the known address).

Reply:
Tried that, but trace[1].op may be ZEND_JIT_TRACE_END immediately, if the code was fully inlined.
```
(gdb) p func->op_array.opcodes
$7 = (zend_op *) 0x748cb840
(gdb) p *trace[1].opline
$8 = {handler = 0x7bff75d0, op1 = {constant = 80, var = 80, num = 80, opline_num = 80, jmp_offset = 80}, op2 = {constant = 4294967184, var = 4294967184, num = 4294967184, opline_num = 4294967184, jmp_offset = 4294967184}, result = {constant = 112, var = 112, num = 112, opline_num = 112, jmp_offset = 112}, extended_value = 0, lineno = 4, opcode = 18 '\022',
  op1_type = 8 '\b', op2_type = 1 '\001', result_type = 18 '\022'}
(gdb) p trace[1].op
$9 = 8 '\b'
```
The normal location for EX(opline) on a function call is `op_array->opcodes + MIN(op_array->num_args, call_num_args)`, which is basically what
```
                                        uint32_t num_args;

                                        if ((func->op_array.fn_flags & ZEND_ACC_HAS_TYPE_HINTS) != 0) {
                                                if (trace) {
                                                        num_args = 0;
                                                } else if (call_info) {
                                                        num_args = skip_valid_arguments(op_array, ssa, call_info);
                                                } else {
                                                        num_args = call_num_args;
                                                }
                                        } else {
                                                num_args = call_num_args;
                                        }
                                        if (zend_accel_in_shm(func->op_array.opcodes)) {
                                                |       LOAD_IP_ADDR (func->op_array.opcodes + num_args)
```
that code computes, i.e. the part enabled by https://github.com/php/php-src/commit/51c74a4fa96ffcd38d3147df9af24b8dcab25635.